pkg/sql/doctor/doctor.go

```diff
@@ -161,6 +162,17 @@ func validateSafely(
 			errs = append(errs, err)
 		}
 	}()
+	// If this is a table, attempt to upgrade its representation so that
+	// validation works as expected.
+	if table, ok := desc.(catalog.TableDescriptor); ok {
+		b := tabledesc.
+			NewBuilderForFKUpgrade(table.TableDesc(), false)
```

Comment:
You might as well be calling `catalogkv.NewBuilder` here, it'll do the same thing. This will allow you to do this for all descriptor types, not just tables, which is a lot more future-proof.

Consequently you might also want to move this logic into `newDescGetter` altogether and upgrade the descriptors at the end of that function, in a second pass because properly upgrading the descriptors with `RunPostDeserializationChanges` requires a valid `DescGetter`. The downside of doing this in `newDescGetter` is that by doing so you effectively lose the "original" descriptors and only ever see their upgraded counterparts thereafter. Not sure if that's important or not, it's up to you.